system/lib/wasmfs/wasmfs.cpp

```diff
@@ -54,20 +54,20 @@ long __syscall_dup(long fd) {
     return -(EBADF);
   }
 
-  return fileTable.add(openFile);
+  return fileTable.add(openFile.unlocked());
 }
 
 __wasi_errno_t __wasi_fd_write(__wasi_fd_t fd,
                                const __wasi_ciovec_t* iovs,
                                size_t iovs_len,
                                __wasi_size_t* nwritten) {
-  std::shared_ptr<OpenFileState> openFile = FileTable::get()[fd];
+  auto openFile = FileTable::get()[fd];
 
   if (!openFile) {
     return __WASI_ERRNO_BADF;
   }
 
-  auto file = openFile->get().getFile()->dynCast<DataFile>();
+  auto file = openFile.locked().getFile()->dynCast<DataFile>();
```

Comment:
I think there is a danger here of the `File` being deleted from another thread, since we're only holding a raw pointer to it and we are also releasing the lock on the `OpenFileState` that is keeping it alive. Maybe as a separate change we should make the `dynCast`, `cast`, and `is` system work with shared pointers rather than raw pointers to ensure liveness of the underlying data.

Reply:
Does line 64 (`auto openFile = FileTable::get()[fd];`) not avoid that? We hold onto OpenFileInfo which holds onto the File.

(Or I guess we have an Entry there, but that holds on to the OpenFileInfo?)

Reply:
>Does line 64 (auto openFile = FileTable::get()[fd];) not avoid that? We hold onto OpenFileInfo which holds onto the File.
(Or I guess we have an Entry there, but that holds on to the OpenFileInfo?)

`Entry` just returns the `OpenFileState`.

Reply:
Hmm yeah, good point @kripken. The `FileTable::Entry` holds the file table lock and we know the file table (indirectly) reaches this file, so the file cannot be deleted while we hold that lock (because the OpenFileState in between cannot be deleted while we hold that lock). It still seems sketchy to be holding raw pointers to `File`s, though :/

Reply:
@ethanalee 

> Entry just returns the OpenFileState but doesn't acquire the OpenFileState.

Oh, that seems dangerous, and to me at least surprising. But I guess it's ok since as @tlively said the Entry locks the table...?

@tlively

I agree we should not have a raw pointer to a File - where does that occur?

Reply:
@kripken Sorry I mean that just holding the Entry doesn't mean you have the `OpenFileState` automatically. You must use `locked()` to `acquire` a locked Handle on `OpenFileState`.

Reply:
Yeah, that would probably be safer if you want to look into it. (Alternatively, I can look into it if you want to focus on more impactful development. Let me know either way.)

Reply:
Sounds good, we can talk about this, I'll make a note of it.